File: golang/set2/challenge9/main.go
package main

import (
	"bytes"
	"errors"
	"fmt"
)

func pkcs7Pad(input []byte, blockSize int) ([]byte, error) {
	if blockSize >= 256 || blockSize <= 0 {
		return nil, errors.New("specified block size is invalid")
	}
	if len(input) > blockSize || len(input) == 0 {
		// XXX should we return an error if input is bigger than block size?
		return nil, errors.New("input block size is invalid")
	}
	padlen := 1
	for ((len(input) + padlen) % blockSize) != 0 {
		padlen = padlen + 1
	}
	pad := bytes.Repeat([]byte{byte(padlen)}, padlen)
	return append(input, pad...), nil
}

func pkcs7Unpad(input []byte, blockSize int) ([]byte, error) {
	if blockSize >= 256 || blockSize <= 0 {
		return nil, errors.New("specified block size is invalid")
	}
	if len(input) > blockSize || len(input) == 0 {
		// XXX should we return an error if input is bigger than block size?
		return nil, errors.New("input block size is invalid")
	}
	padlen := int(input[len(input)-1])
	if padlen > blockSize || padlen == 0 {
		return nil, errors.New("invalid padding")
	}
	pad := input[len(input)-padlen:]
	for i := 0; i < padlen; i++ {
		if pad[i] != byte(padlen) {
			return nil, errors.New("invalid padding")
		}
	}
	return input[:len(input)-padlen], nil
}

func main() {
	input := "YELLOW SUBMARINE"
	padded, err := pkcs7Pad([]byte(input), 20)
	if err != nil {
		panic(err)
	}
	fmt.Printf("padded: %x\n", padded)
	unpadded, err := pkcs7Unpad([]byte(padded), 20)
	if err != nil {
		panic(err)
	}
	if !bytes.Equal([]byte(input), unpadded) {
		panic("padding failure")
	}
}

File: rust/set2/challenge9/src/main.rs
extern crate rustc_serialize as serialize;

use serialize::hex::ToHex;

fn pkcs7_pad(input: &[u8], block_size: u8) -> Vec<u8> {
    let mut pad_len = 1;
    while (input.len() + pad_len) % block_size as usize != 0 {
        pad_len += 1
    }
    let pad = vec![pad_len as u8; pad_len];
    let mut ret: Vec<u8> = vec![];
    ret.extend_from_slice(input);
    ret.extend(pad);
    return ret
}

fn main() {
    let input = "YELLOW SUBMARINE";
    let block = pkcs7_pad(input.as_ref(), 20);
    let out_str = String::from_utf8_lossy(block.as_ref());
    println!("{}", out_str);
    println!("{}", block.to_hex());
}
